datafusion/expr/src/udf.rs

```diff
@@ -249,6 +261,22 @@ pub trait ScalarUDFImpl: Debug + Send + Sync {
     /// the arguments
     fn return_type(&self, arg_types: &[DataType]) -> Result<DataType>;
 
+    /// What [`DataType`] will be returned by this function, given the types of
+    /// the expr arguments
+    fn return_type_from_exprs(
+        &self,
+        arg_exprs: &[Expr],
+        schema: &DFSchema,
+    ) -> Result<DataType> {
```

Comment:
```suggestion
    fn return_type_from_exprs<S: ExprSchema>(
        &self,
        arg_exprs: &[Expr],
        schema: S,
    ) -> Result<DataType> {
```
@Weijun-H If I'd like to make a change like this, there's an error 🥲
> for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically;

Reply:
I am unfamiliar with this part. It seems you need to extend `ScalarUDFImpl` to allow `ExprSchema`.

Reply:
> Maybe here changing return_type to return_type_from_exprs would be a good idea. But a problem
is here the schema has type S: ExprSchema while return_type_from_exprs can't accept a generic type parameter S because it's in a trait. 🤔 What do you think?

Maybe you can use a trait object like `&dyn ExprSchema` 🤔

Reply:
I tried `&dyn ExprSchema` before, there will be an issue when calling `get_type`, like below. 🥲. I also tried the fix method recommended by the complier(adding `+?Sized` but failed)
<img width="1252" alt="Screenshot 2024-01-29 at 9 40 10 AM" src="https://github.com/apache/arrow-datafusion/assets/69566909/0c0a2a8c-f16f-4ec5-bd26-d779b3aa200e">

Reply:
@alamb Sorry for the delay.
I think the only problem that blocks me is the argument type of the function. I had some issue when passing `schema` as a param. 
I can also search for some ways to do it.🤔

Reply:
Cool -- thanks -- I'll try and find some time to play around with it

Reply:
@yyy1000 the following changes worked for me. If anyone else have a better solution please suggest it. I am still a newbie in rust but trying to help to learn. 

expr_schema.rs

<img width="1414" alt="image" src="https://github.com/apache/arrow-datafusion/assets/7233005/04b03f96-4469-4858-a1f1-98d66701fd8b">

<img width="1414" alt="image" src="https://github.com/apache/arrow-datafusion/assets/7233005/62d7e6ed-5c0f-4bd3-b07d-81d12adcb1f9">

return_types_udf.rs

<img width="1414" alt="image" src="https://github.com/apache/arrow-datafusion/assets/7233005/42ab80c7-fc3a-45d4-b4df-ef3843d0a08b">

Reply:
Those look like great changes to me @brayanjuls  -- 👌 

What do you think @yyy1000 ?

Reply:
Thanks for help! @alamb @brayanjuls 
However, I have tried this method before and it will not work.
I will show you the reason on my side. :)

Reply:
I'm not super familiar with `ExprSchema`, but a quick look seems like it's only applicable to a single Expr, and therefore a single column instead of the entire schema?


I think retaining the entire schema would be more flexible. For example, [this](https://github.com/GlareDB/glaredb/issues/2113) is only possible if you have the entire schema.